test/dynamo/test_activation_checkpointing.py

```diff
@@ -1287,6 +1298,213 @@ def fn(x):
 
         self.assertEqual(opt_fn(x), fn(x))
 
+    @requires_distributed()
+    @torch._dynamo.config.patch(inline_inbuilt_nn_modules=True)
+    def test_side_effect_in_forward_hooks(self):
+        """
+        NOTE(yf225): Status:
+        1. [WIP] Running forward pre-hook in compiled AC
+            - Questions: 
+            1. What if the partitioner is not planning to recompute those modules (and instead is planning to save the activations)? What does it mean to "rerun forward hook" for that case?
+            2. Do we really need to rerun forward hook in the backward? What fails if we don't?
+        2. [TODO] Running forward hook in compiled AC
+            - TODO: allow returning tuple from module forward hook
+        3. [TODO] Clean up and deduplicate (can we avoid global BackwardState?)
+
+        NOTE: We need support for both forward pre-hook and forward hook in AC region, because eager FSDP2 uses it
+        (verified by patching https://gist.github.com/yf225/ae73b47c14f56f2b96ea3bfdfe2ed30c and then running test_train_parity_with_activation_checkpointing unit test).
```

Comment:
this is actually not true - those forward hooks are returned immediately upon entry when being run in backward.